### DISCUSSION TITLE
mqtt-over-ws-client with user and password doesn't work

### DISCUSSION THREAD

**Discussion by ondonadas** [cesanta/mongoose#2489] [Ask questions, get answers] [answered]
Hi forum
I'm trying to modify mqtt-over-ws-client example with MQTT user and password fields (no TLS enabled).

In order to achieve that I've filled user/pass before to call mg_mqtt_login.
![image](https://github.com/cesanta/mongoose/assets/1852881/7668c0ab-224d-4cc8-91a3-41590e4f33d3)
Hopefully is the right way as @scaprile mention in [this](https://github.com/cesanta/mongoose/discussions/2488) discussion.

Running a mosquitto eclipse MQTT broker container (localhost) I can't establish the connection. Basically, the websocket connection is established but it mg_mqtt_login doesn't return anything from broker.

> mqtt-over-ws-client (master)$ ./example 
11f266f 3 net.c:171:mg_connect          1 -1 ws://localhost:9001
--- ev=1
--- ev=3
11f266f 3 sock.c:363:mg_connect_resolve 1 4 -> 127.0.0.1:9001 pend
--- ev=4
11f266f 3 sock.c:285:write_conn         1 4 snd 148/2048 rcv 0/0 n=148 err=0
--- ev=8
11f26d4 3 sock.c:274:read_conn          1 4 snd 0/2048 rcv 0/2048 n=129 err=0
--- ev=7
--- ev=11
11f26d4 2 main.c:54:fn                  Connected to WS. Logging in to MQTT...
11f26d4 3 sock.c:285:write_conn         1 4 snd 41/2048 rcv 0/2048 n=41 err=0
11f26d4 2 sock.c:105:iolog              
-- 1 0.0.0.0:0 -> 127.0.0.1:9001 41
0000   82 a3 02 34 94 95 12 15 94 91 4f 65 c0 c1 06 f4   ...4......Oe....
0010   94 95 02 31 fd f1 33 06 a7 95 06 41 e7 f0 70 34   ...1..3....A..p4
0020   9c e5 63 47 e7 e2 6d 46 f0                        ..cG..mF.       
--- ev=8

I've try HIVEMQ and works as expected.
> mqtt-over-ws-client (master)$ ./example 
10dd16c 3 net.c:171:mg_connect          1 -1 ws://broker.hivemq.com:8000/mqtt
--- ev=1
10dd16c 3 net.c:171:mg_connect          2 -1 udp://8.8.8.8:53
10dd16d 3 sock.c:143:mg_send            2 4 0:0 35 err 0
10dd18d 3 sock.c:274:read_conn          2 4 snd 0/0 rcv 0/2048 n=67 err=0
10dd18d 3 dns.c:165:dns_cb              1 broker.hivemq.com is 52.28.39.196
--- ev=3
10dd18d 3 sock.c:363:mg_connect_resolve 1 5 -> 52.28.39.196:8000 pend
--- ev=4
10dd1aa 3 sock.c:285:write_conn         1 5 snd 160/2048 rcv 0/0 n=160 err=0
--- ev=8
10dd1c5 3 sock.c:274:read_conn          1 5 snd 0/2048 rcv 0/2048 n=129 err=0
--- ev=7
--- ev=11
10dd1c5 2 main.c:54:fn                  Connected to WS. Logging in to MQTT...
10dd1c5 3 sock.c:285:write_conn         1 5 snd 41/2048 rcv 0/2048 n=41 err=0
10dd1c5 2 sock.c:105:iolog              
-- 1 0.0.0.0:0 -> 52.28.39.196:8000 41
0000   82 a3 64 47 74 9b 74 66 74 9f 29 16 20 cf 60 87   ..dGt.tft.). .`.
0010   74 9b 64 42 1d ff 55 75 47 9b 60 32 07 fe 16 47   t.dB..UuG.`2...G
0020   7c eb 05 34 07 ec 0b 35 10                        |..4...5.       
--- ev=8
10dd1e0 3 sock.c:274:read_conn          1 5 snd 0/2048 rcv 0/2048 n=6 err=0
10dd1e0 2 sock.c:105:iolog              
-- 1 0.0.0.0:0 <- 52.28.39.196:8000 6
0000   82 04 20 02 00 00                                 .. ...          
--- ev=7
--- ev=12
10dd1e0 2 main.c:77:fn                  GOT 4 bytes WS msg
--- ev=16
10dd1e0 2 main.c:86:fn                  CONNECTED to ws://broker.hivemq.com:8000/mqtt
10dd1e0 2 main.c:93:fn                  SUBSCRIBED to topcic
10dd1e0 3 mqtt.c:326:mg_mqtt_pub        1 [topcic] -> [hello]
10dd1e0 2 main.c:100:fn                 PUBLISHED hello -> topcic
10dd1e1 3 sock.c:285:write_conn         1 5 snd 42/2048 rcv 0/2048 n=42 err=0
10dd1e1 2 sock.c:105:iolog              
-- 1 0.0.0.0:0 -> 52.28.39.196:8000 42
0000   82 8d 7f c4 c2 88 fd cf c2 89 7f c2 b6 e7 0f a7   ................
0010   ab eb 7e 82 91 93 3a a5 1b a1 35 a5 1d e7 55 d5   ..~...:...5...U.
0020   78 fa 59 a5 19 fb 5f c9 77 fc                     x.Y..._.w.      
--- ev=8
10dd1ff 3 sock.c:274:read_conn          1 5 snd 0/2048 rcv 0/2048 n=7 err=0
10dd1ff 2 sock.c:105:iolog              
-- 1 0.0.0.0:0 <- 52.28.39.196:8000 7
0000   82 05 90 03 00 01 01                              .......         
--- ev=7
--- ev=12
10dd1ff 2 main.c:77:fn                  GOT 5 bytes WS msg
10dd20c 3 sock.c:274:read_conn          1 5 snd 0/2048 rcv 0/2048 n=6 err=0
10dd20c 2 sock.c:105:iolog              
-- 1 0.0.0.0:0 <- 52.28.39.196:8000 6
0000   82 04 40 02 00 02                                 ..@...          
--- ev=7
--- ev=12
10dd20c 2 main.c:77:fn                  GOT 4 bytes WS msg
10dd20e 3 sock.c:274:read_conn          1 5 snd 0/2048 rcv 0/2048 n=19 err=0
10dd20e 2 sock.c:105:iolog              
-- 1 0.0.0.0:0 <- 52.28.39.196:8000 19
0000   82 11 32 0f 00 06 74 6f 70 63 69 63 00 33 68 65   ..2...topcic.3he
0010   6c 6c 6f                                          llo             
--- ev=7
--- ev=12
10dd20e 2 main.c:77:fn                  GOT 17 bytes WS msg
10dd20e 3 main.c:109:fn                 1 [topcic] -> [hello]
10dd20e 2 main.c:111:fn                 RECEIVED hello <- topcic
--- ev=9
10dd20e 3 net.c:148:mg_close_conn       1 5 closed
10dd210 3 net.c:148:mg_close_conn       2 4 closed
10dd210 3 net.c:234:mg_mgr_free         All connections closed

If anybody has experience with mqtt over webosocket + login, I've created a [zip file attached](https://github.com/cesanta/mongoose/files/13451896/mqtt-user-pass.zip) containing the modified source code + Wireshark captures (one for hivemq and one with mosquito localhost broker). Thank you.

**Comment by scaprile** [comment]
#2488 #2490 
Please do not open several questions (and issues) for the same subject

**Reply by scaprile** [reply, marked as answer]
Given that HiveMQ works, I don't see this as a Mongoose issue.
In your capture I see the websocket communication seems to establish, but **your broker does not answer**, it is not listening for MQTT in that port. Frame 13 is the login attempt, it is the very same 35-byte content that is sent to HiveMQ as frame 8 in the other capture file:
```text
0000   10 21 00 04 4d 51 54 54 04 c0 00 00 00 05 69 64   .!..MQTT.À....id
0010   31 32 33 00 04 75 73 65 72 00 08 70 61 73 73 77   123..user..passw
0020   6f 72 64                                          ord
```
10: MQTT connect control message, no flags
21: 33, remaining length
00 04: Protocol Name length
MQTT
04: v3.1.1
etc. https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718028
This not a Mongoose issue, "it mg_mqtt_login doesn't return anything from broker" is incorrect, _your_ broker does not return a response to what is a correct Mongoose request issued by your call to mg_mqtt_login.
This is not a problem with mqtt over webosocket + login, you have an issue when configuring your broker to work the way you want it to. If you followed some random tutorial, make sure you are not enabling TLS in your broker, try it with another client, etc.

MOSQUITTO (and other non-port-based broker) USERS:

If your MQTT over WS broker is not port based, that is, it does not assume that protocol carried by the WebSocket connection will be MQTT based on the port number, then, if you can't change it, you'll have to add an extra header in the connection request, in order to negotiate that:
`Sec-Websocket-Protocol: mqtt`
So, change this line:
https://github.com/cesanta/mongoose/blob/30738f6638dea27be4f8e423136227618c57c44e/examples/mqtt-over-ws-client/main.c#L101
to this:
`mg_ws_connect(&mgr, s_url, fn, &done, "%s", "Sec-Websocket-Protocol: mqtt\r\n");`

**Comment by scaprile** [comment]
I can't guess with no network captures. I've seen your capture where Mongoose sends and Mosquitto does not answer...
I don't know your network nor your configurations, and that mix of what looks like IPv6 addresses... your Mosquitto is saying it is answering, but there is no answer in your capture, so if it is answering... then to whom ?
The only thing I can suggest is that you try with some other public broker somewhere else.
AFAICS, our code works, this is no Mongoose issue.
Look me up and PM me in Spanish (email preferred), I'll try to help you just because of your name, but offline, I don't think this belongs here.
